cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll

```diff
@@ -63,6 +63,12 @@ class Expr extends StmtParent, @expr {
    * order of destruction.
    */
   DestructorCall getImplicitDestructorCall(int n) {
+    exists(Expr e |
+      e = this.(TemporaryObjectExpr).getExpr() and
+      synthetic_destructor_call(e, max(int i | synthetic_destructor_call(e, i, _)) - n, result)
+    )
+    or
+    not this = any(TemporaryObjectExpr temp).getExpr() and
```

Comment:
@jketema I'm curious to know if this is something you think we could fix in the extractor: the problem is that the destructor is attached to the expression instead of it's associated `TemporaryObjectExpr`. So when we generate IR for such an expression `e` we get:
```
// ir for e
// destructor call
// ir for the TemporaryObjectExpr
```
and the problem is that the destructor call mentions the temporary which is only defined by the IR for the `TemporaryObjectExpr`.

So if we could instead attach the destructor call to the `TemporaryObjectExpr` then we wouldn't need to move these destructor calls around at the QL level.

I'm okay with this being done at the QL level since it seems to work. But if just as easy to attach it to the `TemporaryObjectExpr` in the extractor as what we're currently doing we might as well do it there. What do you think?

Reply:
Could you point me to some relevant PrintAST output? It's very difficult for me to follow along without that 😅

Reply:
Understandable 😂 You can see an example in the commit leading up to the fix here: https://github.com/github/codeql/pull/16125/commits/a0de95dc4428bad6fda8576516f7a0f8274d2d15#diff-e8b861918b5d430185cc157d26a66c7e8e038d34685c395a57e2565797184b55R2285-R2296. I've highlighted the problem in the link: the destructor call is attached to `call to make` and thus comes before the `TemporaryObjectExpr`. And since the qualifier of the destructor call is a reuse of `TemporaryObjectExpr` (which hasn't "happened" yet) we get problematic IR.

You can see the effect on PrintAST in this diff: https://github.com/github/codeql/compare/89eaadd76fd7170a0292cf5355780d0b6321c804..d40fa4cfba0b81b43273e450ddd5714b29c55ead (which includes the changes I had to do to `PrintAST` to get the destructor call back in the PrintAST output because we didn't handle destructor calls attached to conversions before).

Reply:
I think this might be possible, yes. My reservation is that you'll lose the explicit information on the order in which the destructions need to happen (because the destructors for different temporaries will no longer be attached in one fixed spot, but instead spread throughout).

Reply:
That's a good point. I think it's worth discussing this with the rest of the team once this PR has landed, I guess.

Reply:
I'd also like to see a at least one test case that has multiple temporaries in a single expression that need destruction, and see the effect https://github.com/github/codeql/compare/89eaadd76fd7170a0292cf5355780d0b6321c804..d40fa4cfba0b81b43273e450ddd5714b29c55ead#diff-e8b861918b5d430185cc157d26a66c7e8e038d34685c395a57e2565797184b55 has on that.

Reply:
I've added one in https://github.com/github/codeql/pull/16125/commits/febd06063aa401cb8790e0363f08c94ff2cb8f7d. Is that kinda what you had in mind? I'm sure there are multiple ways of creating such a situation, but this was the first I could think of 😅

Reply:
Yeah that's what I had in mind, and that also looks odd. They're all attached to the "last" temporary. That's something that'll very difficult to replicate in the extractor without doing extra traversals over the AST to figure out what that last temporary actually is.

Reply:
I also wonder what happens when the top-level expression does not have a temporary. Where are the destructors attached in that case?

Reply:
Will do. I've also opened https://github.com/github/codeql/pull/16146 to add some tests for it.

Reply:
I've force-pushed your suggested example in https://github.com/github/codeql/pull/16125/commits/9c25ce4079cf91d0026b771243eed41fd0781238. I think both the IR and the PrintAST looks correct, although there's still a temporary expression for the qualifier to `getClassWithDestructor2`.

Reply:
Yeah, that's what I would expect. It's not a the top-level of the expression, which was the important thing.

Reply:
> So alternatively you might be able to some something like that, but for xvalues and for `getParent` and `getChild`, although that'll be a bit of an invasive change because you'll need to change the `getChild` of `Expr` (and the `getParent` of `ReuseExpr`).

This might not be completely correct, because the synthetic destructors are hooked up via `expr_ancestor` and not via `exprparents`. So if you want to go this route, you'll have to adapt wherever `expr_ancestor`is used.

Reply:
Thanks for the heads up. If you're okay with this as-is I'd rather keep it like this this local change seems to do the job for now.

Reply:
That's perfectly fine with me.